parsing/extensions_parsing.ml

```diff
@@ -311,94 +307,32 @@ end)
-      ('ast, 'ext_ast) optional_ast_extension
-  | Unsupported
-
-module Translate
-         (Syntactic_category : sig
-            type ('ast, 'ext_ast) t
-
-            val ast_module :
-              ('ast, 'ext_ast) t ->
-              (module AST with type ast = 'ast)
-
-            val ast_extension :
-              ('ast, 'ext_ast) t ->
-              Clflags.Extension.t ->
-              ('ast, 'ext_ast) optional_ast_extension
-          end) : sig
-  val extension_ast_of_ast :
-    ('ast, 'ext_ast) Syntactic_category.t ->
-    'ast ->
-    'ext_ast option
-
-  val ast_of_extension_ast :
-    ('ast, 'ext_ast) Syntactic_category.t ->
-    loc:Location.t ->
-    Clflags.Extension.t ->
-    'ext_ast ->
-    'ast
+module type Of_ast_parameters = sig
+  module AST : AST
+  type t
+  val of_ast_internal : Clflags.Extension.t -> AST.ast -> t option
+end
+
+module Make_of_ast (Params : Of_ast_parameters) : sig
+  val of_ast : Params.AST.ast -> Params.t option
 end = struct
-  let extension_ast_of_ast
-        (type ast ext_ast)
-        (cat : (ast, ext_ast) Syntactic_category.t)
-        (ast : ast)
-      : ext_ast option =
-    let (module AST) = Syntactic_category.ast_module cat in
-    let loc = AST.location ast in
+  let of_ast ast =
+    let loc = Params.AST.location ast in
     let raise_error err = raise (Error (loc, err)) in
-    match AST.match_extension ast with
+    match Params.AST.match_extension ast with
     | None -> None
     | Some ([name], ast) -> begin
         match Clflags.Extension.of_string name with
         | Some ext -> begin
             assert_extension_enabled ~loc ext;
-            match Syntactic_category.ast_extension cat ext with
-            | Supported { of_ast; wrap; _ } ->
-                Some (wrap (of_ast ast))
-            | Unsupported ->
-                raise_error (Wrong_syntactic_category(ext, AST.plural))
+            match Params.of_ast_internal ext ast with
+            | Some ext_ast -> Some ext_ast
+            | None -> raise_error (Wrong_syntactic_category(ext, Params.AST.plural))
           end
         | None -> raise_error (Unknown_extension name)
       end
     | Some ([], _) ->
         raise_error Unnamed_extension
     | Some (name :: subnames, _) ->
         raise_error (Bad_introduction(name, subnames))
-
-  let ast_of_extension_ast
-        (type ast ext_ast)
-        (cat : (ast, ext_ast) Syntactic_category.t)
-        ~(loc : Location.t)
-        (extn : Clflags.Extension.t)
-        (east : ext_ast)
-      : ast =
-    let (module AST) = Syntactic_category.ast_module cat in
-    let raise_error err = raise (Error(loc, err)) in
-    match Syntactic_category.ast_extension cat extn with
-    | Supported { unwrap; ast_of; _ } -> begin
-        assert_extension_enabled ~loc extn;
-        match unwrap east with
-        | Some east' ->
-            AST.make_extension
-              ~loc
-              [Clflags.Extension.to_string extn]
-              (ast_of ~loc east')
```

Comment:
The loss of this implicit node is my biggest sadness around this change; it's the one place we gain boilerplate from what you wrote instead of lose it.  Do you have any thoughts on workarounds?  The issue seems to be that when we fragmented the `{Expression,Pattern}.ast_of` into `Some_extension.{expr,pat}_of`, this was the only piece we couldn't share.  Do you have any thoughts?  I was considering perhaps a function parameterized by a module, so something like `Expression.ast_of (module Immutable_array)`, but I'm not convinced.

Reply:
I will resolve this conversion in favor of the other one.